lib/galaxy/webapps/galaxy/api/jobs.py

```diff
@@ -230,19 +231,13 @@ def __get_job( self, trans, id ):
             decoded_job_id = self.decode_id( id )
         except Exception:
             raise exceptions.MalformedId()
-        query = trans.sa_session.query( trans.app.model.Job )
-        if trans.user_is_admin():
-            query = query.filter(
-                trans.app.model.Job.id == decoded_job_id
-            )
-        else:
-            query = query.filter(
-                trans.app.model.Job.user == trans.user,
-                trans.app.model.Job.id == decoded_job_id
-            )
-        job = query.first()
+        job = trans.sa_session.query( trans.app.model.Job ).filter( trans.app.model.Job.id == decoded_job_id ).first()
         if job is None:
             raise exceptions.ObjectNotFound()
+        if not trans.user_is_admin() and job.user != trans.user:
+            for data_assoc in job.output_datasets:
+                if not self.dataset_manager.is_accessible( data_assoc.dataset.dataset, trans.user ):
+                    raise exceptions.ItemAccessibilityException( "You are not allowed to rerun this job." )
```

Comment:
No worries. This is actually very important. So I tested this earlier and it worked. I assumed that it works because its validated on the basis of the dataset itself and not the hda.